### PR TITLE
Deployment pipeline for dev

### PR DESCRIPTION
**Pull request by idolgov**
# Description

- Enable (automatic) deployment to dev environment upon merging to master
- Add proper linting and testing to the CI

## Why was the change made?

We need to deploy to dev environment.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Unittest
- [x] Cypress e2e -tests

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/index.html [hunk 1]
```diff
@@ -10,15 +10,15 @@
     <title>Ylitse MentorApp - SOS-Lapsikylä</title>
     <link rel="stylesheet"
       type="text/css"
-      href="/static/styles/fonts.css" />
+      href="styles/fonts.css" />
     <link rel="stylesheet"
       type="text/css"
-      href="/static/styles/normalize.css" />
+      href="styles/normalize.css" />
```
Comment (kuosmark) [inline src/index.html:16]: Does this work? 🤔

File: src/login/index.html [hunk 2]
```diff
@@ -11,14 +11,14 @@
...
       content="width=device-width, initial-scale=1" />
     <link rel="stylesheet"
-      href="/static/styles/normalize.css" />
+      href="../styles/normalize.css" />
     <link rel="stylesheet"
-      href="/static/styles/static-pages.css" />
+      href="../styles/static-pages.css" />
     <link rel="stylesheet"
-      href="/login/styles.css" />
+      href="styles.css" />
```
Comment (kuosmark) [inline src/login/index.html:18]: Nice one!

File: src/login/index.html [hunk 3]
```diff
@@ -35,14 +35,14 @@
         </div>
         <div class="panels">
           <div id="mobile-header">
-            <img src="/static/img/ylitse-logo-tiny.svg">
+            <img src="../img/ylitse-logo-tiny.svg">
```
Comment (kuosmark) [inline src/login/index.html:38]: Does this really work?

**Review by kuosmark — COMMENTED** [review]
Can you explain using the copy-webpack plugin to me? Why do we copy the contents of /register, /login and /static? This is also probably why the e2e tests don't work. How is this fixed?

Also: Could we be running the e2e tests in the CI pipeline against a clean db?

**Review by rottabonus — COMMENTED** [review]
As a user, I had to make this change to make the logout work correctly:
file: /src/features/Logout/Logout.tsx
```diff
const Logout = () => {
       try {
         const response = await fetch('/api/logout');
         if (response.ok) {
-          window.location.href = '/login';
+          window.location.href = '/login/';
         }
       } catch (error) {
         console.error(error);
```

The e2e tests:
- Changing the paths for the test-runner visit-commands in a similar way that I suggested above (i.e /register => /register/) also made the styles to load correctly as before they were styleleless. If you would visit /login, the styles would be missing, but /login/ gets the styles
- The login-test always tries to open the register-page for some reason, and so the test fails

And on another note.. When these work better, it would be nice to include **npm run lint** on the pipeline as a test-job

**Review by rottabonus — APPROVED** [review]
Very nice job!

I would approve this now.
I got some issues when running the e2e-suite locally (in every suite some tests), but the pros of having these checks in pipeline + path to automatic deployment weights over these.

If you could add a ticket for fixing the flaky tests, then I would definitely merge this

File: package.json [hunk 4]
```diff
@@ -4,15 +4,16 @@
...
-    "lint": "npx tsc && npm run eslint && npm run prettier"
+    "eslint": "eslint .",
+    "lint": "tsc && npm run eslint:check && npm run eslint",
+    "test:unit": "jest",
+    "test:e2e": "npm run wait && cypress run",
+    "test": "npm run test:unit && npm run test:e2e",
+    "build": "webpack",
+    "start": "webpack serve --open",
+    "wait": "wait-on http://localhost:8080/version && wait-on http://localhost:8082"
   },
```
Comment (kuosmark) [inline package.json:17]: Very nice improvement!

File: src/login/index.html [hunk 5]
```diff
@@ -62,19 +62,26 @@ <h1 localization-key="loginTitle"></h1>
                 placeholder="" />
               <a href="#"
                 localization-key="retrievePasswordLink"></a>
+              <label for="mfa_token"
+                localization-key="mfaLabel"></label>
+              <input id="mfa_token"
+                name="mfa_token"
```
Comment (kuosmark) [inline src/login/index.html:68]: Just a minor nitpick, but the other IDs and names are named using hyphens so this could also be **mfa-token**.

**Review by kuosmark — APPROVED** [review]
Great stuff! I'm really glad we finally have a working CI/CD pipeline!

File: src/login/index.html (hunk 5 above)
Comment (idolgov) [reply src/login/index.html:68]: No problem, I'll change it to hyphens :+1:

**Review by kuosmark — APPROVED** [review]

**Review by rottabonus — DISMISSED** [review]

**Review by kuosmark — CHANGES_REQUESTED** [review]
One quick observation!

**Review by rottabonus — APPROVED** [review]
asd?

**Comment by idolgov** [comment]
Github bug?

**Pull request merged**